nix/overlays/isabelle.nix

```diff
@@ -0,0 +1,41 @@
+self: super:
+let
+  pkgs = super;
+
+  pkgsForOldZ3 = import (builtins.fetchTarball {
+    name = "nixos-unstable-z3-old";
+    url = "https://github.com/nixos/nixpkgs/archive/da6eec001ab9437a50d4f91d8c1a3704c193647c.tar.gz";
+    sha256 = "0rzxhi45rxxqw44wf0hvyby993a0gjq2kzygq27rjp11wn71q5kf";
+  }) {};
+  z3old = pkgsForOldZ3.z3;
+
+  pkgsForOldPolyML = import (builtins.fetchTarball {
+    name = "nixos-unstable-polyml-old";
+    url = "https://github.com/nixos/nixpkgs/archive/c381bdc1c542a885bea1ac2209f4dc6449206127.tar.gz";
+    sha256 = "0zwhq69hzhcvnigmvxlkqv8nq825s2m6iwikb7fjsj096kr8q2ds";
+  }) {};
+  polymlPatched = pkgsForOldPolyML.polyml.overrideAttrs(args: pkgs.recurseIntoAttrs(rec {
+    configureFlags = args.configureFlags ++ ["--enable-intinf-as-int"];
+  }));
+
+  pkgsForIsabelle = import (builtins.fetchTarball {
+    name = "nixos-unstable-polyml-old";
+    url = "https://github.com/nixos/nixpkgs/archive/5272327b81ed355bbed5659b8d303cf2979b6953.tar.gz";
+    sha256 = "0182ys095dfx02vl2a20j1hz92dx3mfgz2a6fhn31bqlp1wa8hlq";
+  }) {};
+
+  isabelle2018 = pkgsForIsabelle.isabelle.override(args: {z3 = z3old; polyml = polymlPatched; });
+
+  isabelle2018bootstrap = isabelle2018.overrideAttrs(args: pkgs.recurseIntoAttrs(rec {
+    installPhase = args.installPhase + ''
+        sed -i 's/RunCall.clearMutableBit state;//g' $out/Isabelle2018/src/Pure/Concurrent/synchronized.ML        
+        #patchShebangs $out
+        #$out/bin/isabelle build -s -b HOL''; # cache HOL
```

Comment:
could you remove these commented out lines if they are not useful, if they are then add a comment explaining.

Reply:
@shmish111 Isabelle initial setup requires HOL library to be built (if it's not - it would rebuild it) and it takes a while. At first I expected this build step to be part of derivation in order to save build time.

 However the tricky thing here is that it puts compiled HOL into user folder '~/.isabelle`, so this step wouldn't speed anything up if derivation is cached.
Nevertheless, I'm considering to use '-s' option that would put HOL into Isabelle installation, so it would be packed ahead of time.

I kinda kept those lines as a reminder in case if build is too slow on CI.